internal/jobqueue/fsjobqueue/fsjobqueue.go

```diff
@@ -668,3 +668,27 @@ func jobMatchesCriteria(j *job, acceptedJobTypes []string, acceptedChannels []st
 
 	return contains(acceptedJobTypes, j.Type) && contains(acceptedChannels, j.Channel)
 }
+
+// AllRootJobIDs Return a list of all the top level(root) job uuids
+// This only includes jobs without any Dependents set
+func (q *fsJobQueue) AllRootJobIDs() ([]uuid.UUID, error) {
+	ids, err := q.db.List()
+	if err != nil {
+		return nil, err
+	}
+
+	jobIDs := []uuid.UUID{}
+	for _, id := range ids {
+		var j job
+		exists, err := q.db.Read(id, &j)
+		if err != nil {
+			return jobIDs, err
+		}
+		if !exists || len(j.Dependents) > 0 {
```

Comment:
I've always considered the root node to be the first job in a chain. So in a depsolve > manifest > osbuild chain it's the depsolve job that's the root I thought, and in a koji job koji-init. The osbuild / koji-finalize would be leaf nodes then.

But this considers the osbuild or koji-finalize job to be the root node, no? As those are the only ones without dependents.

Reply:
I haven't seen any live SQL examples, but with the fsjobqueue backend when you start a compose the first job is the one with no Dependents. TBF I find the terminology here hard to keep straight in my head, but from what I see happening Dependencies are jobs the current job is a parent of, and Dependents are jobs that depend on the current job.